File: src/Test/RBTreeFFI.fs
module Test.RBTreeFFI
type Color = Red | Black
type Tree = Empty | Node of Color * Tree * int * Tree
let balance = function
    | Black, Node(Red, Node(Red, a, x, b), y, c), z, d
    | Black, Node(Red, a, x, Node(Red, b, y, c)), z, d
    | Black, a, x, Node(Red, Node(Red, b, y, c), z, d)
    | Black, a, x, Node(Red, b, y, Node(Red, c, z, d)) -> Node(Red, Node(Black, a, x, b), y, Node(Black, c, z, d))
    | c, a, x, b -> Node(c, a, x, b)
let insert x t =
    let rec ins = function
        | Empty -> Node(Red, Empty, x, Empty)
        | Node(c, a, y, b) ->
            if x < y then balance(c, ins a, y, b)
            elif x > y then balance(c, a, y, ins b)
            else Node(c, a, y, b)
    match ins t with
    | Node(_, a, y, b) -> Node(Black, a, y, b)
    | Empty -> Empty
let rec count = function Empty -> 0 | Node(_, a, _, b) -> 1 + count a + count b
let runRBTreeFFI (n: obj) =
    let rec build i t = if i > (unbox<int> n) then t else build (i + 1) (insert i t)
    count (build 1 Empty) :> obj

File: src/Test/RBTreeFFI.php
<?php
$exports['runRBTreeFFI'] = function($limit) {
    $empty = (object)['tag' => 'E'];
    $node = function($color, $left, $value, $right) {
        return (object)['tag' => 'T', 'color' => $color, 'left' => $left, 'value' => $value, 'right' => $right];
    };
    $red = function($tree) { return $tree->tag === 'T' && $tree->color === 'R'; };
    $balance = function($color, $a, $x, $b) use ($node, $red) {
        if ($color === 'B') {
            if ($red($a) && $red($a->left)) {
                $child = $a->left;
                return $node('R', $node('B', $child->left, $child->value, $child->right), $a->value, $node('B', $a->right, $x, $b));
            }
            if ($red($a) && $red($a->right)) {
                $child = $a->right;
                return $node('R', $node('B', $a->left, $a->value, $child->left), $child->value, $node('B', $child->right, $x, $b));
            }
            if ($red($b) && $red($b->left)) {
                $child = $b->left;
                return $node('R', $node('B', $a, $x, $child->left), $child->value, $node('B', $child->right, $b->value, $b->right));
            }
            if ($red($b) && $red($b->right)) {
                $child = $b->right;
                return $node('R', $node('B', $a, $x, $b->left), $b->value, $node('B', $child->left, $child->value, $child->right));
            }
        }
        return $node($color, $a, $x, $b);
    };
    $ins = function($value, $tree) use (&$ins, $balance, $node, $empty) {
        if ($tree->tag === 'E') return $node('R', $empty, $value, $empty);
        if ($value < $tree->value) return $balance($tree->color, $ins($value, $tree->left), $tree->value, $tree->right);
        if ($value > $tree->value) return $balance($tree->color, $tree->left, $tree->value, $ins($value, $tree->right));
        return $node($tree->color, $tree->left, $tree->value, $tree->right);
    };
    $depth = function($tree) use (&$depth) {
        return $tree->tag === 'E' ? 0 : 1 + max($depth($tree->left), $depth($tree->right));
    };
    $tree = $empty;
    for ($value = (int)$limit; $value > 0; $value--) {
        $redRoot = $ins($value, $tree);
        $tree = $node('B', $redRoot->left, $redRoot->value, $redRoot->right);
    }
    return $depth($tree);
};
return $exports;
